### PR TITLE
Fix PEM certificate loading that sometimes fails

### PR DESCRIPTION
**Pull request by blenk92**
As described in https://github.com/openssl/openssl/issues/9187, the
loading of PEM certificates sometimes fails if a line of base64
content has the length of a multiple of 254.
The problem is in get_header_and_data(). When such a line with a
length of 254 (or a multiple) has been read, the next read will
only read a newline. Due to this get_header_and_data() expects to be
in the header not in the data area. This commit fixes that by checking
if lines have been read completely or only partially. In case of a
previous partial read, a newline will be ignored.

Created a new PR as adviced during submitting (I|C)CLA
Old PR: https://github.com/openssl/openssl/pull/11484

##### Checklist

- [x ] tests are added or updated

### CODE REVIEW

File: crypto/pem/pem_lib.c [hunk 1]
```diff
@@ -828,6 +828,12 @@ static int get_header_and_data(BIO *bp, BIO **header, BIO **data, char *name,
             goto err;
         }
 
+        /* Check if line has been read completely or if only part of the line
+         * has been read. Keep the previous value to ignore newlines that
+         * appear due to reading a line up until the char before the newline */
```
Comment (t8m) [inline crypto/pem/pem_lib.c:833]: Style nit - the end of multiline comment should be on a separate line with the `*` aligned.

File: crypto/pem/pem_lib.c [hunk 2]
```diff
@@ -838,13 +844,17 @@ static int get_header_and_data(BIO *bp, BIO **header, BIO **data, char *name,
 
         /* Check for end of header. */
         if (linebuf[0] == '\n') {
-            if (got_header == POST_HEADER) {
-                /* Another blank line is an error. */
-                PEMerr(PEM_F_GET_HEADER_AND_DATA, PEM_R_BAD_END_LINE);
-                goto err;
+            /* If previous line has been read only partially this newline is a
+             * regular newline at the end of a line and not an empty line */
```
Comment (t8m) [inline crypto/pem/pem_lib.c:848]: As above.

File: crypto/pem/pem_lib.c (hunk 1 above)
Comment (blenk92) [reply crypto/pem/pem_lib.c]: okay, done

File: crypto/pem/pem_lib.c (hunk 2 above)
Comment (blenk92) [reply crypto/pem/pem_lib.c]: done, as well ;-)

File: crypto/pem/pem_lib.c (hunk 1 above)
Comment (kaduk) [reply crypto/pem/pem_lib.c]: Shouldn't the opening `/*` also be on a line of its own (for a multiline comment)?

File: crypto/pem/pem_lib.c [hunk 3]
```diff
@@ -828,6 +828,13 @@ static int get_header_and_data(BIO *bp, BIO **header, BIO **data, char *name,
             goto err;
         }
 
+        /* Check if line has been read completely or if only part of the line
+         * has been read. Keep the previous value to ignore newlines that
+         * appear due to reading a line up until the char before the newline
```
Comment (kaduk) [inline crypto/pem/pem_lib.c:833]: If you make other changes (e.g., to the formatting of the start of the comment), please also add a period to end this sentence.

**Comment by blenk92** [comment]
> Thanks again for putting this together with the extra test cases -- it looks good (modulo the comment style for which @t8m and I need to check what is right)

okay, thanks for the fast reviews :-)

let me know if something needs to be changed...

**Comment by kaduk** [comment]
Chapter 8 at https://www.openssl.org/policies/codingstyle.html says that the opening `/*` should be on its own line.  I don't think that these comments have particular formatting, so the `-` in `/*-` is not needed.  @t8m please confirm

**Comment by blenk92** [comment]
> Chapter 8 at https://www.openssl.org/policies/codingstyle.html says that the opening `/*` should be on its own line. I don't think that these comments have particular formatting, so the `-` in `/*-` is not needed. @t8m please confirm

Alright then, i guess if it's written there it makes sense to change it...

**Comment by kaduk** [comment]
Thanks; I reconfirm it looks good to me.

**Comment by t8m** [comment]
Yeah, I forgot about that rule. Still approved after the comment style change.

**Comment by kaduk** [comment]
It's been more than 24 hours since @t8m reconfirmed, so I'll go ahead and start merging this one.  I will keep @beldmit as a reviewer given the trivial nature of the changes since his approval

**Comment by kaduk** [comment]
Merged manually to master and 1.1.1, so closing.
Thank you again for the fix!!

**Pull request closed without merging**